### ISSUE TITLE
after_failure output gets truncated unless sleep is used

### ISSUE THREAD

**Issue by Malvineous** [travis-ci/travis-ci#6018]
If you have an `after_failure` section that uses `cat` to display more detail about the failure for example, then only just over one line make it into the console log and later commands do not appear at all.  It seems the VM is terminated too early, before all the output has been fully captured.

By adding a `sleep 10` after the `cat` command, this seems to keep the VM around long enough to ensure all the output is captured, but it seems a bit dodgy to me.  It looks like the VM is being terminated as soon as the process finishes, without waiting for stdout/stderr to finish flushing first.

Here is an example of the following `.travis.yml`:

```
after_failure:
  - if [ -f tests/tests.log ]; then cat tests/tests.log; fi
  - echo == End of test log ==
```

This [gets cut off and the final `echo` never appears](https://travis-ci.org/Malvineous/libgamearchive/jobs/128519222):

```
$ if [ -f tests/tests.log ]; then cat tests/tests.log; fi
Running 1182 test cases...
unkn
```

However adding a delay like this:

```
after_failure:
  - if [ -f tests/tests.log ]; then cat tests/tests.log && sleep 10; fi
  - echo == End of test log ==
```

Correctly [captures all the output](https://travis-ci.org/Malvineous/libgamearchive/jobs/128574599):

```
$ if [ -f tests/tests.log ]; then cat tests/tests.log && sleep 10; fi
Running 1182 test cases...
unknown location:0: fatal error: in "test_bnk_harry/test_archive::test_insert_end[bnk_harry]": memory access violation at address: 0x00000000: no mapping at fault address
test-archive.cpp:163: last checkpoint
*** 1 failure is detected in the test module "libgamearchive"
FAIL tests (exit status: 201)

$ echo == End of test log ==
== End of test log ==

Done. Your build exited with 1.
```

This seems to mainly be a problem under the OSX VMs.

**Comment by CleanCut** [comment]
I am having this same behavior in `before_install`.  Thanks for the tip to add a sleep command.  That helped me get my actual failure output.

**Comment by tueda** [comment]
A similar thing happened to me even for Linux builds (for example, [a truncated log](https://travis-ci.org/tueda/form/jobs/248291790) vs. [a completed log for a similar job](https://travis-ci.org/tueda/form/jobs/248291791)). Adding `sleep 1` to `after_script` resolves the problem so far.

**Comment by edigiacomo** [comment]
I solved this problem limiting the rate of the output with `pv -q -L 3k` (https://github.com/ARPA-SIMC/eccodes-rpm/blob/9f94a7a497e6f4c2252480670ad1de68c3b446a8/.travis-build.sh#L38)

**Comment by native-api** [comment]
Note that [only `stdout` seems affected, not `stderr`](https://github.com/travis-ci/travis-ci/issues/9055).

**Comment by mmdriley** [comment]
I think I know what's going on here. FWIW, this should only be happening in builds with secret environment variables.

Travis uses [this script](https://github.com/travis-ci/travis-build/blob/0cb66e041c67e2e717f6ebec8050c43ace74a0d1/public/filter.rb) to replace secret values with `[secure]`. In one mode of operation, that script takes `stdin` and filters it to `stdout`.

In what Travis calls the `redirect_io` filter strategy, [Bash's stdout is replaced](https://github.com/travis-ci/travis-build/blob/f952a16e511c6cd59e641d24c522671fb668db63/lib/travis/build/appliances/setup_filter.rb#L55-L62) with a [process substitution](http://tldp.org/LDP/abs/html/process-sub.html) that runs through `filter.rb` and thence to `stdout`.

I made a build with secret environment variables and printed the build script. You can see the redirect code in `travis_run_setup_filter` [here](https://travis-ci.org/mmdriley/travis-playground/builds/348122898#L859-L878).

```bash
export TRAVIS_FILTERED=redirect_io
exec 9>&1 1> >(
  
  curl -sf -o ~/filter.rb https://build.travis-ci.org/filter/redirect_io.rb
  if [ $? -ne 0 ]; then
    echo "Download from https://build.travis-ci.org/filter/redirect_io.rb failed. Trying https://build.travis-ci.com/filter/redirect_io.rb ..."
    curl -sf -o ~/filter.rb https://build.travis-ci.com/filter/redirect_io.rb
  fi

  export SECRET_0=[secure]
  ruby ~/filter.rb -e "SECRET_0"
) 2>&1
```

Except process substitution is subject to race conditions! If you run `cmd1 > >(cmd2)` then the shell does not wait for `cmd2` to exit before reporting an exit code for `cmd1`. For example:

```
$ true > >(sleep 1; echo done); echo $?
0
$ done
```

See [this StackOverflow answer](https://stackoverflow.com/a/4493552/8872982) or [this thread on `comp.unix.shell`](https://groups.google.com/d/msg/comp.unix.shell/GqLNzUA4ulA/SmpAzEubF7gJ).

You can create a 100% repro by sending the Ruby process `SIGSTOP` -- the build output is truncated from there, but the build still completes successfully because the runner sees exit status 0. Example [here](https://travis-ci.org/mmdriley/travis-playground/builds/348363267).

It seems Travis knows this is a problem and has an [alternative filtering strategy called `pty`](https://github.com/travis-ci/travis-build/blob/f952a16e511c6cd59e641d24c522671fb668db63/lib/travis/build/appliances/setup_filter.rb#L47-L54) that they're [rolling out incrementally](https://github.com/travis-ci/travis-build/blob/f952a16e511c6cd59e641d24c522671fb668db63/lib/travis/build/appliances/setup_filter.rb#L87-L89) (see [this commit](https://github.com/travis-ci/travis-build/commit/8bd81eac750b67fa09f016471964872e7ecc9bd1)). The `pty` strategy makes `ruby` the parent of the running command, so `ruby` has a chance to see all output before exiting.

**Comment by vinriviere** [comment]
This bug is extremely annoying.
I also use secret variables, and the log output is always truncated.

**Comment by mmdriley** [comment]
@vinriviere you -- or anyone else experiencing this problem -- should be able to contact Travis support (at `support@travis-ci.com`) and ask them to move your projects over to the `pty` strategy. As far as I can tell, this lever is the only way to consistently solve this problem and is only accessible to Travis staff.

**Comment by samupl** [comment]
The issue is still appearing, are there any fixes right now, or is contacting travis to move to pty the only way?

**Comment by Totktonada** [comment]
I tried the workaround with `sleep 10` in `after_script` and it does not work for me: I saw a truncated output despite it.

I asked Travis-CI support to enable pty filtering strategy for https://github.com/tarantool/tarantool-php, but the answer was:

> Unfortunately, we can only enable this per GitHub username, not for the GitHub organization name.

Since my request was about one repository, not the entire organization, I guess enabling it for one repository within a GitHub organization is not possible too.

No resolution for me so. I asked Travis-CI support about plans to fix this issue and waiting for an answer now. I'll update the issue with the answer if Travis-CI staff will not share plans here before this.
